app/Actions/ServicePayment/PayAction.php

```
@@ -0,0 +1,90 @@
+<?php
+namespace App\Actions\ServicePayment;
+
+use App\Exceptions\InsufficientFundsException;
+use App\Models\Service;
+use App\Models\ServicePaymentRequest;
+use App\Models\User;
+use App\Models\Wallet;
+use App\Models\WalletTransaction;
+use App\Traits\FormatApiResponse;
+use App\Traits\TransactionTrait;
+use Exception;
+use Illuminate\Support\Facades\DB;
+use Throwable;
+
+class PayAction
+{
+    use FormatApiResponse, TransactionTrait;
+
+   /**
+    *
+    * @param array $data
+    * @return JsonResponse
+    */
+    public function execute(User $user, array $data)
+    { 
+        DB::beginTransaction();
+        try{
+
+            $service = Service::getServiceClass($data['service_name']);
+
+            $fee = 0;
+
+            $reference = $this->generateReference(Wallet::LABEL);
+            
+            $smartCardNumber = isset($data['smartcard_number']) ? $data['smartcard_number'] : null;
+            $customerDetails = $service->getSmartCardDetails($smartCardNumber);
+
+            if($data['subscription_type'] == 'renew' && isset($customerDetails['renewal_amount'])) {
+                $amount = $customerDetails['renewal_amount'];
+                $data['amount'] = $amount;
+            } else{
+                $variations = $service->getVariations();
+
+                $variation = array_filter($variations, function ($item) use ($data) {
+                    return $item['variation_code'] == $data['variation_code'];
+                });
+    
+                $variation = reset($variation); 
+
+                if ($variation) {
+                    $amount = $variation['amount'];
+                    $data['variation_amount'] = $variation['amount'];
+                } else {
+                    throw new Exception("Variation not found.");
+                }
+            }
+
+            $user->wallet->debit($reference, $amount, $fee, WalletTransaction::TYPE['WITHDRAW'], ucfirst($data['service_name']).' Service Payment');
```

I just realised this part of the code will fail current users that don't have wallet. Can we throw a clear error message that says that the payment failed because the user is yet to add money to their wallet. I.e. We can treat it as insufficient balance.

---

@PinkCode15

---

Everyone would have a wallet, there's a command to generate wallet for existing users